### PR TITLE
dpdk: rework hugepage hints to use per-numa information v2

### PR DESCRIPTION
**Pull request by lukashino**
Follow-up #9735

Previous integration of hugepage analysis only fetched data from /proc/meminfo. However, this proved to be often deceiving mainly for providing only global information and not taking into account different hugepage sizes (e.g. 1GB hugepages) and different NUMA nodes.

Describe changes:
- removed guards as much as I could
- but the guards are needed to avoid compilation errors - I tried to remove them and it failed - see https://github.com/lukashino/suricata/actions/runs/7004361732/job/19052076585
- moved DPDK-specific logic to suricata.c only
- replaced dirent.h with suricata.h

https://redmine.openinfosecfoundation.org/issues/6419

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 16735

**Review by victorjulien — CHANGES_REQUESTED** [review]
cppcheck
```
src/util-hugepages.c:120:9: warning: %hd in format string (no. 1) requires 'short' but the argument type is 'unsigned short'. [invalidPrintfArgType_sint]
        snprintf(path, sizeof(path),
        ^
src/util-hugepages.c:129:13: warning: %hd in format string (no. 1) requires 'short *' but the argument type is 'unsigned short *'. [invalidScanfArgType_int]
        if (fscanf(f, "%hd", &hugepages[i].allocated) != 1) {
            ^
src/util-hugepages.c:137:9: warning: %hd in format string (no. 1) requires 'short' but the argument type is 'unsigned short'. [invalidPrintfArgType_sint]
        snprintf(path, sizeof(path),
        ^
src/util-hugepages.c:146:13: warning: %hd in format string (no. 1) requires 'short *' but the argument type is 'unsigned short *'. [invalidScanfArgType_int]
        if (fscanf(f, "%hd", &hugepages[i].free) != 1) {
            ^
src/util-hugepages.c:126:20: warning: Either the condition '!f' is redundant or there is possible null pointer dereference: f. [nullPointerRedundantCheck]
            fclose(f);
                   ^
src/util-hugepages.c:124:13: note: Assuming that condition '!f' is not redundant
        if (!f) {
            ^
src/util-hugepages.c:126:20: note: Null pointer dereference
            fclose(f);
                   ^
src/util-hugepages.c:143:20: warning: Either the condition '!f' is redundant or there is possible null pointer dereference: f. [nullPointerRedundantCheck]
            fclose(f);
                   ^
src/util-hugepages.c:141:13: note: Assuming that condition '!f' is not redundant
        if (!f) {
            ^
src/util-hugepages.c:143:20: note: Null pointer dereference
            fclose(f);
                   ^
src/util-hugepages.c:126:20: error: Null pointer dereference [nullPointer]
            fclose(f);
                   ^
src/util-hugepages.c:143:20: error: Null pointer dereference [nullPointer]
            fclose(f);
                   ^
```
scan-build about the same
```
In function 'SystemHugepagePerNodeGetHugepageInfoLinux',
    inlined from 'SystemHugepagePerNodeGetHugepageInfo' at util-hugepages.c:178:11,
    inlined from 'SystemHugepageSnapshotCreate' at util-hugepages.c:300:23:
util-hugepages.c:126:13: warning: argument 1 null where non-null expected [-Wnonnull]
  126 |             fclose(f);
      |             ^~~~~~~~~
In file included from suricata-common.h:64,
                 from suricata.h:67,
                 from util-hugepages.c:24:
/usr/include/stdio.h: In function 'SystemHugepageSnapshotCreate':
/usr/include/stdio.h:183:12: note: in a call to function 'fclose' declared 'nonnull'
  183 | extern int fclose (FILE *__stream) __nonnull ((1));
      |            ^~~~~~
In function 'SystemHugepagePerNodeGetHugepageInfoLinux',
    inlined from 'SystemHugepagePerNodeGetHugepageInfo' at util-hugepages.c:178:11,
    inlined from 'SystemHugepageSnapshotCreate' at util-hugepages.c:300:23:
util-hugepages.c:143:13: warning: argument 1 null where non-null expected [-Wnonnull]
  143 |             fclose(f);
      |             ^~~~~~~~~
/usr/include/stdio.h: In function 'SystemHugepageSnapshotCreate':
/usr/include/stdio.h:183:12: note: in a call to function 'fclose' declared 'nonnull'
  183 | extern int fclose (FILE *__stream) __nonnull ((1));
      |            ^~~~~~
util-hugepages.c:126:13: warning: Null pointer passed to 1st parameter expecting 'nonnull' [core.NonNullParamChecker]
            fclose(f);
            ^~~~~~~~~
util-hugepages.c:143:13: warning: Null pointer passed to 1st parameter expecting 'nonnull' [core.NonNullParamChecker]
            fclose(f);
            ^~~~~~~~~
2 warnings generated.
```

File: src/util-hugepages.c [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        return;
+
+    for (uint16_t i = 0; i < s->num_nodes; i++) {
+        SystemHugepageNodeInfoDestroy(&(s->nodes[i]));
+    }
+    SCFree(s->nodes);
+    SCFree(s);
+}
+
+SystemHugepageSnapshot *SystemHugepageSnapshotCreate(void)
```
[the remark this answers is not in the document]
Comment (lukashino) [reply src/util-hugepages.c:275]: added

File: src/util-hugepages.c [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+    uint16_t node_cnt = SystemNodeCountGet();
+    if (node_cnt == 0) {
+        SCLogError("failed to obtain number of NUMA nodes in the system");
+        return NULL;
+    }
+    NodeInfo *nodes = SCCalloc(node_cnt, sizeof(*nodes));
+    if (nodes == NULL) {
+        FatalError("failed to allocate memory for NUMA node info");
+        return NULL;
```
[the remark this answers is not in the document]
Comment (lukashino) [reply src/util-hugepages.c:288]: return statement removed

**Comment by lukashino** [comment]
Continues in https://github.com/OISF/suricata/pull/10164

**Pull request closed without merging**